source/common/http/http2/codec_impl.cc

```diff
@@ -835,51 +843,58 @@ void ConnectionImpl::sendPendingFrames() {
   }
 }
 
-void ConnectionImpl::sendSettings(const Http2Settings& http2_settings, bool disable_push) {
-  ASSERT(http2_settings.hpack_table_size_ <= Http2Settings::MAX_HPACK_TABLE_SIZE);
-  ASSERT(Http2Settings::MIN_MAX_CONCURRENT_STREAMS <= http2_settings.max_concurrent_streams_ &&
-         http2_settings.max_concurrent_streams_ <= Http2Settings::MAX_MAX_CONCURRENT_STREAMS);
-  ASSERT(
-      Http2Settings::MIN_INITIAL_STREAM_WINDOW_SIZE <= http2_settings.initial_stream_window_size_ &&
-      http2_settings.initial_stream_window_size_ <= Http2Settings::MAX_INITIAL_STREAM_WINDOW_SIZE);
-  ASSERT(Http2Settings::MIN_INITIAL_CONNECTION_WINDOW_SIZE <=
-             http2_settings.initial_connection_window_size_ &&
-         http2_settings.initial_connection_window_size_ <=
-             Http2Settings::MAX_INITIAL_CONNECTION_WINDOW_SIZE);
-
-  std::vector<nghttp2_settings_entry> iv;
-
-  if (http2_settings.allow_connect_) {
-    iv.push_back({NGHTTP2_SETTINGS_ENABLE_CONNECT_PROTOCOL, 1});
-  }
-
-  if (http2_settings.hpack_table_size_ != NGHTTP2_DEFAULT_HEADER_TABLE_SIZE) {
-    iv.push_back({NGHTTP2_SETTINGS_HEADER_TABLE_SIZE, http2_settings.hpack_table_size_});
-    ENVOY_CONN_LOG(debug, "setting HPACK table size to {}", connection_,
-                   http2_settings.hpack_table_size_);
-  }
-
-  if (http2_settings.max_concurrent_streams_ != NGHTTP2_INITIAL_MAX_CONCURRENT_STREAMS) {
-    iv.push_back({NGHTTP2_SETTINGS_MAX_CONCURRENT_STREAMS, http2_settings.max_concurrent_streams_});
-    ENVOY_CONN_LOG(debug, "setting max concurrent streams to {}", connection_,
-                   http2_settings.max_concurrent_streams_);
-  }
-
-  if (http2_settings.initial_stream_window_size_ != NGHTTP2_INITIAL_WINDOW_SIZE) {
-    iv.push_back(
-        {NGHTTP2_SETTINGS_INITIAL_WINDOW_SIZE, http2_settings.initial_stream_window_size_});
-    ENVOY_CONN_LOG(debug, "setting stream-level initial window size to {}", connection_,
-                   http2_settings.initial_stream_window_size_);
+void ConnectionImpl::sendSettings(
+    const envoy::config::core::v3::Http2ProtocolOptions& http2_options, bool disable_push) {
+  std::unordered_set<nghttp2_settings_entry, SettingsEntryHash, SettingsEntryEquals> settings;
+  // Custom parameters override named parameters, so they must be inserted first into the set.
```

Comment:
> If they override, should we either force-disable server push first, or reject that custom setting?

Thanks for catching this, it does require special handling given that it is disabled for server connections but enabled for client connections. Unfortunately, this creates awkwardness around error handling when an end user attempts to enable it via a custom parameter, as Envoy claims not to support server push but does end up sending the parameter to downstream peers (even though it never actually sends PUSH_PROMISE frames).

Do you have context around why it is enabled for downstream clients?

> Alternately it might be worth considering it a config error to set a field twice instead of debug logging. WDYT?

The intent behind the current logic is to simplify forward compatibility with future named parameters. If an end user is setting a custom parameter (e.g., 0x20) that Envoy eventually ends up supporting as a named parameter, I would rather avoid breaking those configs when the user upgrades Envoy given that this new field addition would not be considered a breaking change to the API under current guidelines.

Reply:
Also no context for why this is enabled for downstream.

@mattklein123 and @goaway any thoughts, or can we just fix this?

Reply:
I understand the idea behind trying to support existing configs when an Envoy upgrade adopts a named parameter, but I wonder if that could lead to hard-to-detect configuration bugs across clients/severs. My first reaction is that I'm somewhat in favor of it just being a configuration error, to aid with earlier detection of a potential conflict.

Reply:
Send what downstream? disable_push is false for server connections?

Reply:
(But yes I think it was done that way just because of a common function, feel free to fix as needed)